wasm: move compilation to an alien thread

wmitros opened the pull request:
The compilation of wasm UDFs is performed by a call to a foreign
function, which cannot be divided with yielding points and, as a
result, causes long reactor stalls for big UDFs.
We avoid them by submitting the compilation task to a non-seastar
std::thread, and retrieving the result using seastar::alien.

The thread is created at the start of the program. It executes
tasks from a queue in an infinite loop.

All seastar shards reference the thread through a std::shared_ptr
to a `alien_thread_runner`.

Considering that the compilation takes a long time anyway, the
alien_thread_runner is implemented with focus on simplicity more
than on performance. The tasks are stored in an std::queue, reading
and writing to it is synchronized using an std::mutex for reading/
writing to the queue, and an std::condition_variable waiting until
the queue has elements.

When the destructor of the alien runner is called, an std::nullopt
sentinel is pushed to the queue, and after all remaining tasks are
finished and the sentinel is read, the thread finishes.

Fixes #12904

wmitros commented:
This approach is still not perfect, I'm particularly concerned about how we're accessing memory between shards (in particular, the final compiled module is allocated on the alien thread, and we're saving the result on each shard).
This should still be an improvement over the first approach at #12904 (the compilation wasn't actually performed on an alien thread there), so I'm posting this as a PR.

avikivity commented on cql3/statements/create_aggregate_statement.cc:
```diff
@@ -22,7 +22,7 @@ namespace cql3 {
 namespace statements {
 
-shared_ptr<db::functions::function> create_aggregate_statement::create(query_processor& qp, db::functions::function* old) const {
+seastar::future<shared_ptr<db::functions::function>> create_aggregate_statement::create(query_processor& qp, db::functions::function* old) const {
```
When converting non-future to future there is always a risk of introducing races - now two such statements can execute concurrently.

We can assume Raft, so if Raft already serializes execution, we're fine. I don't have any concrete reasons to think  there's a race here, but I'm not too familiar with the code path, so please double check it.

wmitros replied:
I examined the new futures again, but couldn't find any races that could be caused by them. A potential race that looked like it might appear was between replacing and dropping the UDF, but it looks like we handle it well enough already.

avikivity commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,96 @@
+namespace wasm {
+
+std::optional<wasm_compile_task> task_queue::pop_front() {
+    if (_sem.try_acquire_for(10ms)) {
```
This just generates unnecessary wakeups. Better to use a condition variable and a mutex, then you can wake up the thread for when something is queued or when _stopped is set. Another option is to have a queue of std::optionals, and stop when nullopt is queued.

tchaikov replied:
yeah, the reason why `ShardedWorkQueue` uses `try_acquire_for()` is get it waked up for checking if the thread pool is stopped or not. in our case, an `std::optional<>` or `std::unique_ptr<>` would serve the purpose better. as Avi puts, the nullopt or an empty pointer would be a `sentinel`. i'd suggest use an `optional<>`, as `task_queue::pop_front()` already returns an `optional<>`.

wmitros replied:
Ok, I tried to use the approach with a condition variable and an `optional` sentinel in the rebase

avikivity commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,96 @@
+            seastar::alien::run_on(alien, shard, [&p, mod = std::move(mod)] () mutable {
+                p.set_value(std::move(mod));
+            });
+        } catch (const rust::Error& e) {
```
Should handle any exception.

wmitros replied:
Ok

avikivity commented on main.cc:
```diff
@@ -1006,13 +1007,19 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             static sharded<wasm::instance_cache> wasm_instance_cache;
             auto udf_enabled = cfg->enable_user_defined_functions() && cfg->check_experimental(db::experimental_features_t::feature::UDF);
             std::any stop_udf_cache_handlers;
+            static std::shared_ptr<wasm::alien_thread_runner> alien_runner = nullptr;
```
You don't need the static if you have correct stop/start. The statics you see are due to bugs.

tchaikov replied:
nit, no need to initialize the `shared_ptr` with nullptr. it's already empty if created with the default ctor.

wmitros replied:
Ok

avikivity commented on main.cc:
```diff
@@ -1146,6 +1153,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             if (udf_enabled) {
                 qp.invoke_on_all([&] (cql3::query_processor& qp) {
                     qp.set_wasm_instance_cache(&wasm_instance_cache.local());
+                    qp.set_alien_runner(alien_runner);
```
What prevents use of the alien runner before it is set?

wmitros replied:
The code structure - similarly to the instance cache - the runner's only used for initialization when parsing system tables (which happens later in the scylla_main()) and when executing queries (which is started even later in scylla_main()).
We could add the runner(and cache) to the query_processor constructor, but the wasm parameters are sill optional (UDFs are experimental) and it's already taking a lot of parameters so I don't know if we want it. We could also add some checks when trying to use them (alien runner and cache), but it seems like it's simpler to just keep the order of operations correct.

avikivity replied:
I think it's nicer to pass it as a constructor parameter, it eliminates the question about use-before-set. It's an idiom in C++, pass your dependencies up-front.

UDF will not be experimental forever (in fact, I hope we can move them out of experimental soon).

wmitros replied:
I did not add this in the last rebase, but I assume we would like to do this also for the instance cache and also wasmtime::Engine (which is currently in the `database`)? If so, maybe I'll just make a follow-up, which moves them all to the constructor

avikivity replied:
ok for a follow-up.

avikivity commented:
Looks good. I had some minor comments.

It would be good to renice the thread so it doesn't compete with reactor threads or the networking core. See pthread_setthreadprio(). If it uses the same priority levels as nice(), then 10 would be reasonable.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4686/

tchaikov commented on test/boost/wasm_alloc_test.cc:
```diff
@@ -98,6 +99,7 @@ SEASTAR_TEST_CASE(test_allocation_failures) {
         // Check that we can actually throw errors both during compilation and during execution
         BOOST_CHECK(errors_during_compilation > 0);
         BOOST_CHECK(errors_during_execution > 0);
+        alien_runner.stop();
```
nit, might need to use `auto stop_alien_runner = deferred_stop(alien_runner);` right after it is defined? more idiomatic this way.

wmitros replied:
Looks like the `deferred_stop` only works with `stop()` methods that return a future, so it can't be used easily here, but this stops being a problem anyway in the new approach.

tchaikov commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,96 @@
+{ }
+
+alien_thread_runner::~alien_thread_runner() {
+    thread.join();
```
how about `assert(pending_queue.size() == 0`? if we don't call `stop()` in the dtor, we might need to make sure the user of `alien_thread_runner` class does this. or, we can do so in the dtor of `task_queue`.

wmitros replied:
I moved the stop to the dtor in the rebase

tchaikov commented on lang/wasm_alien_thread_runner.hh:
```diff
@@ -0,0 +1,57 @@
+class alien_thread_runner {
+    std::thread thread;
+    bool _stopping = false;
+    task_queue pending_queue;
```
might want to add `_` prefix to the private member variables to be more consistent with the rest of this project.

wmitros replied:
Yes, I'm surprised I missed that

wmitros commented on lang/wasm.cc:
```diff
@@ -7,6 +7,7 @@
  */
 
 #include "wasm.hh"
+#include "seastar/coroutine/exception.hh"
```
Ok

wmitros commented:
I fixed some comments in the rebase (the answered ones). 

Regarding the remaining comments, I think we should maybe consider handling the stopping of the alien runner less gently. In my manual tests I saw that we finish all compilations before stopping the runner anyway (even if they take a long time), so the stop() method doesn't really do anything other than preventing further tasks from being added, which probably shouldn't happen if we we're already requested to stop.

One more issue that I noticed with this PR is that now, when all compilations are performed on one thread, for each UDF we perform the same compilation again and again for each shard. Instead, we could compile it once and cache it, and then copy the result for the following calls. The runtime that we're using actually has a feature of caching modules, but it uses `mmap` for allocating memory for them, so I disabled them by default. We can consider using, or just implement our our cache.
Either way, we can probably do it as a follow up.

wmitros commented:
I addressed the review comments in the rebase. The stopping code is now only present in the destructor of the `alien_thread_runner`. @avikivity @tchaikov please take a look

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4751/

avikivity commented on lang/wasm_alien_thread_runner.hh:
```diff
@@ -0,0 +1,52 @@
+    std::mutex _mut;
+    std::condition_variable _cv;
+    static constexpr unsigned QUEUE_SIZE = 128;
+    std::queue<std::optional<wasm_compile_task>> _pending;
```
I can see this morphing to be a unique_ptr<some_base_class_type> later on, but no need to change it now.

avikivity commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,75 @@
+    std::unique_lock<std::mutex> lock(_mut);
+    if (work_item && _pending.size() >= QUEUE_SIZE) {
+        seastar::alien::run_on(seastar::engine().alien(), work_item->shard, [&done = work_item->done] {
+            done.set_exception(std::runtime_error("alien_thread_runner queue is full"));
```
Could this cause use to fail to load?

Consider 150 shards trying to compile a UDF in parallel. The last 30 will fail. How do we recover from this?

Better to have an unbounded queue.

wmitros replied:
Ok

avikivity commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,75 @@
+}
+
+void alien_thread_runner::submit(seastar::promise<rust::Box<wasmtime::Module>>& p, std::function<rust::Box<wasmtime::Module>()> f) {
+    std::function<void()> packaged([f=std::move(f), &p, &alien = seastar::engine().alien(), shard = seastar::this_shard_id()] () mutable {
```
spaces around `=`.

wmitros replied:
Ok

avikivity commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,75 @@
+            seastar::alien::run_on(alien, shard, [&p, mod = std::move(mod)] () mutable {
+                p.set_value(std::move(mod));
+            });
+        } catch (const std::exception& e) {
```
`catch (...)`

wmitros replied:
Ok

tchaikov commented on lang/wasm.cc:
```diff
@@ -227,8 +235,12 @@ void precompile(context& ctx, const std::vector<sstring>& arg_names, std::string
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
     }
+    co_return;
```
nit, no need to call `co_return` explicitly here.

wmitros replied:
Ok

tchaikov commented on lang/wasm_alien_thread_runner.cc:
```diff
@@ -0,0 +1,75 @@
+namespace wasm {
+
+std::optional<wasm_compile_task> task_queue::pop_front() {
+    std::unique_lock<std::mutex> lock(_mut);
```
```suggestion
    std::unique_lock lock(_mut);
```

as the compiler is able to deduce the template parameter.

wmitros replied:
Ok

tchaikov commented on lang/wasm_alien_thread_runner.hh:
```diff
@@ -0,0 +1,52 @@
+
+class alien_thread_runner {
+    std::thread _thread;
+    task_queue _pending_queue;
```
probably need to put `_pending_queue` before `_thread`, so `_pending_queue` is initialized before `_thread`, as in the constructor of `alien_thread_runner`, it initializes `_thread` using `_pending_queue`.

avikivity replied:
Good catch.

tchaikov commented on lang/wasm_alien_thread_runner.hh:
```diff
@@ -0,0 +1,52 @@
+struct wasm_compile_task {
+    std::function<void()> func;
+    seastar::promise<rust::Box<wasmtime::Module>>& done;
+    int shard;
```
might want to use `unsigned`. to be consistent with the function signature of `seastar::alien::run_on()` . FWIW, we have `using shard_id = unsigned;`, and use `unsigned` and `shard_id` interchangeably.   but, in general, we don't use a signed int for shard.

wmitros replied:
Makes sense

wmitros commented on configure.py:
```diff
@@ -1015,6 +1015,7 @@ def find_headers(repodir, excluded_dirs):
                 'mutation_writer/feed_writers.cc',
                 'lang/lua.cc',
                 'lang/wasm.cc',
+                'lang/wasm_alien_thread_runner.cc',
```
Sure. I think the cmake build was not yet synced with the `lang` directory when I first posted the PR, but now the cmake should be kept up to date with each new PR

tchaikov replied:
yeah, it's been under active development. this file was actually added in last week.

wmitros commented:
I rebased on top of master and addressed the last review comments (except [one](https://github.com/scylladb/scylladb/pull/13051#discussion_r1127898722))

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4786/

tchaikov commented on lang/alien_task_queue.hh:
```diff
@@ -0,0 +1,16 @@
+    class alien_task_queue {
+        
+    }
+}
```
is this file a leftover?

wmitros replied:
Yes, sorry for that

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4787/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4792/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4803/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4809/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4816/

tchaikov commented on lang/wasm_alien_thread_runner.hh:
```diff
@@ -0,0 +1,52 @@
+struct task_queue {
+    std::mutex _mut;
+    std::condition_variable _cv;
+    static constexpr unsigned QUEUE_SIZE = 128;
```
it seems we don't use this variable anymore after switching to unbounded queue and using the combination of `mutex` and `condition_variable` for coordinating the consumer and producer of the queue.

avikivity replied:
btw, an unbounded queue has its own obvious problems. But I think it cannot grow in an unbounded way because Raft will serialize schema changes.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4832/

wmitros commented:
@avikivity I addressed all suggestions (maybe apart from https://github.com/scylladb/scylladb/pull/13051#discussion_r1127898722) and we got one LGTM, do you have any more comments?
